toolkit/tools/pkg/licensecheck/licensecheck.go

```diff
@@ -0,0 +1,452 @@
+// - rpmPath: The path to the RPM to check relative to the chroot's root.
+func CheckRpmLicenses(rpmPath, distTag string, exceptions LicenseExceptions) (result LicenseCheckResult, err error) {
+	defines := rpm.DefaultDistroDefines(false, distTag)
+
+	_, files, _, documentFiles, licenseFiles, err := rpm.QueryPackageContents(rpmPath, defines)
+	if err != nil {
+		return LicenseCheckResult{RpmPath: rpmPath, err: fmt.Errorf("failed to query package contents. Error:\n%w", err)}, nil
+	}
+
+	pkgNameLines, err := rpm.QueryPackage(rpmPath, "%{NAME}", defines)
+	if err != nil {
+		return LicenseCheckResult{RpmPath: rpmPath, err: fmt.Errorf("failed to query package. Error:\n%w", err)}, nil
+	}
+	if len(pkgNameLines) != 1 {
+		return LicenseCheckResult{RpmPath: rpmPath, err: fmt.Errorf("failed to query package. Error: expected 1 package name, got %d", len(pkgNameLines))}, nil
+	}
+	pkgName := pkgNameLines[0]
+
+	badDocFiles, badOtherFiles, duplicatedDocs := interpretResults(pkgName, files, documentFiles, licenseFiles, exceptions)
+
+	result = LicenseCheckResult{
+		RpmPath:        rpmPath,
+		PackageName:    pkgName,
+		BadDocs:        badDocFiles,
+		BadFiles:       badOtherFiles,
+		DuplicatedDocs: duplicatedDocs,
+	}
+
+	return result, nil
+}
+
+// interpretResults scans file lists for packing issues:
+// - badDocFiles: %doc files that appear to be licenses, but are not at least also in the license files
+// - badOtherFiles: files that are misplaced in the licenses directory
+// - duplicatedDocs: %doc files that are also in the license files
+func interpretResults(pkgName string, files, documentFiles, licenseFiles []string, exceptions LicenseExceptions) (badDocFiles, badOtherFiles, duplicatedDocs []string) {
+	badDocFiles = []string{}
+	badOtherFiles = []string{}
+	duplicatedDocs = []string{}
+
+	// Check the documentation files
+	for _, documentFile := range documentFiles {
+		if IsALicenseFile(pkgName, documentFile) && !exceptions.ShouldIgnoreFile(pkgName, documentFile) {
+			if isDocumentInLicenseFiles(documentFile, licenseFiles) {
+				duplicatedDocs = append(duplicatedDocs, documentFile)
+			} else {
+				badDocFiles = append(badDocFiles, documentFile)
+			}
+		}
+	}
+
+	// Make sure we don't put random files in the license directory. They need to be %license.
+	licenseFileSet := sliceutils.SliceToSet(licenseFiles)
+	for _, file := range files {
+		if isFileMisplacedInLicensesFolder(file, licenseFileSet) && !exceptions.ShouldIgnoreFile(pkgName, file) {
+			badOtherFiles = append(badOtherFiles, file)
+		}
+	}
+
+	sort.Strings(badDocFiles)
+	sort.Strings(duplicatedDocs)
+	sort.Strings(badOtherFiles)
+
+	return badDocFiles, badOtherFiles, duplicatedDocs
+}
+
+// isDocumentInLicenseFiles checks if a document file is in the list of license files (based on basename of the file).
+func isDocumentInLicenseFiles(documentFile string, licenseFiles []string) bool {
```

Comment:
Why are you only checking this filename, instead of the full path?

Reply:
Many packages will do something like:
```
%install
cp LICENSE.txt %{buidlroot}/some/random/folder/LICENSE.txt
...
%files
%license LICENSE.txt
%doc some/random/folder/LICENSE.txt
```

I couldn't think of a better way to try and detect that duplication. It will likely lead to some false-positive hits but we can use exceptions if needed. I think in general for a  well behaved package (ie no files like LICENSE.txt in %doc) this becomes a non-issue (since there are no docs to check in the first place).